pkg/agent/cniserver/pod_monitor.go

```diff
@@ -0,0 +1,246 @@
+	maxRetryDelay = 300 * time.Second
+)
+
+var (
+	// retryInterval is the interval to re-install Pod OpenFlow entries if any error happened.
+	// Note, using a variable rather than constant for retryInterval because we may use a shorter time in the
+	// test code.
+	retryInterval = time.Second * 5
+)
+
+type unreadyPodInfo struct {
+	podName        string
+	podNamespace   string
+	flowsInstalled bool
+	createTime     time.Time
+}
+
+type podIfaceMonitor struct {
+	podLister         corelisters.PodLister
+	kubeClient        clientset.Interface
+	eventBroadcaster  record.EventBroadcaster
+	record            record.EventRecorder
+	ifaceStore        interfacestore.InterfaceStore
+	ofClient          openflow.Client
+	podUpdateNotifier channel.Notifier
+
+	// unreadyInterfaces is a map to store the OVS ports which is waiting for the PortStatus from OpenFlow switch.
+	// The key in the map is the OVS port name, and its value is unreadyPodInfo.
+	// It is used only on Windows now.
+	unreadyInterfaces map[string]*unreadyPodInfo
+	// unreadyMapMutex is a mutex to guard map unreadyInterfaces.
+	unreadyMapMutex sync.Mutex
+	statusCh        chan *openflow15.PortStatus
+
+	queue workqueue.TypedRateLimitingInterface[string]
+	// statusMapMtx is a mutex to guard ovsPortStatusMap
+	statusMapMtx     sync.Mutex
+	ovsPortStatusMap map[string]uint32
+}
+
+func (m *podIfaceMonitor) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting the monitor to wait for new OpenFlow ports")
+	m.eventBroadcaster.StartStructuredLogging(0)
+	m.eventBroadcaster.StartRecordingToSink(&v1.EventSinkImpl{
+		Interface: m.kubeClient.CoreV1().Events(""),
+	})
+	defer m.eventBroadcaster.Shutdown()
+
+	go wait.Until(m.worker, time.Second, stopCh)
+
+	go wait.Until(func() {
+		ticker := time.Tick(retryInterval)
+		for {
+			select {
+			case status := <-m.statusCh:
+				klog.V(2).InfoS("Received PortStatus message", "message", status)
+				// Update Pod OpenFlow entries only after the OpenFlow port state is live.
+				m.processPortStatusMessage(status)
+			case <-ticker:
+				m.checkUnreadyPods()
+			}
+		}
+	}, time.Second, stopCh)
+}
+
+func (m *podIfaceMonitor) processPortStatusMessage(status *openflow15.PortStatus) {
+	// Update Pod OpenFlow entries only after the OpenFlow port state is live.
+	if status.Desc.State == openflow15.PS_LIVE {
+		ovsPort := string(bytes.Trim(status.Desc.Name, "\x00"))
+		ofPort := status.Desc.PortNo
+		m.statusMapMtx.Lock()
+		m.ovsPortStatusMap[ovsPort] = ofPort
+		m.statusMapMtx.Unlock()
+		m.queue.Add(ovsPort)
+	}
+}
+
+func (m *podIfaceMonitor) processNextWorkItem() bool {
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+	defer m.queue.Done(key)
+
+	if err := m.updateUnreadyPod(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		m.queue.Forget(key)
+
+		m.statusMapMtx.Lock()
+		defer m.statusMapMtx.Unlock()
+		delete(m.ovsPortStatusMap, key)
```

Comment:
this should probably be done in `updateUnreadyPod`, and this function should only be about queue management, not business logic

Reply:
Sure, updated.